File: Backend/Filmer-Backend/src/main/java/com/filmer/filmerbackend/Services/FriendsService.java
package com.filmer.filmerbackend.Services;

import com.filmer.filmerbackend.Entities.Users;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
/**
 * Interfejs serwisu do zarządzania operacjami związanymi z listą znajomych i profilami użytkowników.
 */
public interface FriendsService {

    /**
     * Wyszukuje użytkowników na podstawie części ich pseudonimu.
     *
     * @param nick część pseudonimu użytkownika
     * @return lista użytkowników spełniających kryteria wyszukiwania
     */
    List<Users> searchUsersByNick(String nick);

    /**
     * Pobiera listę znajomych dla danego użytkownika.
     *
     * @param userId identyfikator użytkownika
     * @return lista znajomych powiązanych z użytkownikiem
     */
    List<Users> getFriendsByUserId(int userId);

    /**
     * Dodaje użytkownika do listy znajomych innego użytkownika.
     *
     * @param userId   identyfikator użytkownika, który dodaje znajomego
     * @param friendId identyfikator dodawanego znajomego
     */
    void addFriend(int userId, int friendId);

    /**
     * Przesyła zdjęcie profilowe dla danego użytkownika.
     *
     * @param userId identyfikator użytkownika
     * @param file   plik zawierający zdjęcie profilowe
     */
    void uploadProfilePicture(int userId, MultipartFile file);

    /**
     * Pobiera zdjęcie profilowe danego użytkownika.
     *
     * @param userId identyfikator użytkownika
     * @return tablica bajtów reprezentująca zdjęcie profilowe
     */
    byte[] getProfilePicture(int userId);

    /**
     * Wysyła zaproszenie e-mail do znajomego.
     *
     * @param friendId  identyfikator znajomego
     * @param lobbyLink link do lobby, które ma być przesłane w zaproszeniu
     */
    void sendInviteEmail(int friendId, String lobbyLink);
}

File: Frontend/Filmer-Frontend/src/app/services/friends.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

/**
 * Serwis FriendsService odpowiedzialny za obsługę listy znajomych,
 * zarządzanie zdjęciem profilowym, wyszukiwanie użytkowników oraz wysyłanie zaproszeń.
 * @export
 * @class FriendsService
 */
@Injectable({
  providedIn: 'root'
})
export class FriendsService {
  /**
   * URL API do obsługi znajomych.
   * @private
   * @type {string}
   */
  private apiUrl = 'http://localhost:8080/friends/api';

  /**
   * Tworzy instancję FriendsService.
   * @param {HttpClient} http Serwis do wykonywania żądań HTTP.
   */
  constructor(private http: HttpClient) {}

  /**
   * Pobiera listę znajomych dla danego użytkownika.
   * @param {number} userId ID użytkownika.
   * @returns {Observable<any[]>} Observable zawierający listę znajomych.
   */
  getFriends(userId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${userId}/list`);
  }

  /**
   * Przesyła zdjęcie profilowe użytkownika na serwer.
   * @param {number} userId ID użytkownika.
   * @param {FormData} formData Obiekt FormData zawierający zdjęcie profilowe.
   * @returns {Observable<any>} Observable z odpowiedzią serwera.
   */
  uploadProfilePicture(userId: number, formData: FormData): Observable<any> {
    return this.http.post(`${this.apiUrl}/${userId}/uploadProfilePicture`, formData);
  }

  /**
   * Pobiera zdjęcie profilowe użytkownika z serwera.
   * @param {number} userId ID użytkownika.
   * @returns {Observable<Blob>} Observable zawierający zdjęcie profilowe jako Blob.
   */
  getProfilePicture(userId: number): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/${userId}/profilePicture`, { responseType: 'blob' });
  }

  /**
   * Wyszukuje użytkowników na podstawie nicku.
   * @param {string} nick Nick użytkownika do wyszukania.
   * @returns {Observable<any[]>} Observable zawierający listę pasujących użytkowników.
   */
  searchUsers(nick: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/search?nick=${nick}`);
  }

  /**
   * Dodaje użytkownika do listy znajomych.
   * @param {number} userId ID użytkownika dodającego znajomego.
   * @param {number} friendId ID użytkownika, który ma zostać dodany jako znajomy.
   * @returns {Observable<void>} Observable z odpowiedzią serwera.
   */
  addFriend(userId: number, friendId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${userId}/addFriend/${friendId}`, {});
  }

  /**
   * Wysyła e-mail z zaproszeniem do lobby do znajomego.
   * @param {number} friendId ID znajomego, do którego ma zostać wysłane zaproszenie.
   * @param {string} lobbyLink Link do lobby.
   * @returns {Observable<void>} Observable z odpowiedzią serwera.
   */
  sendInviteEmail(friendId: number, lobbyLink: string): Observable<void> {
    console.log({ friendId, lobbyLink });
    return this.http.post<void>(`${this.apiUrl}/sendInvite`, { friendId, lobbyLink });
  }
}
